### PR TITLE
Shell meta-sequence tokens

### LINKED ISSUE #16: Implement subshell grammar
https://github.com/mgattozzi/ysh/blob/9d6a58d1d5aebd561789c094766ef6605fb4c8c7/src/parse.rs#L175-L183

### PR DESCRIPTION
**Pull request by myrrlyn**
Implement shell meta-sequence tokenizing. Closes #16

This PR depends on #29. Any changes to #29 will require modifying this branch to take those changes into account.

### CODE REVIEW

**Comment by hawkw** [comment]
It looks like this needs to be rebased onto master before it will merge cleanly --- @myrrlyn can I get you to do that?

**Comment by hawkw** [comment]
Unclear why the appveyor branch build is still failing if you rebased this onto master. It's Probably Fine.

File: src/token.rs [hunk 1]
```diff
@@ -279,6 +291,106 @@ pub fn keyval(text: &str) -> TokenResult<(&str, &str)> {
...
 }
 
+/// Finds a shell meta-sequence.
+///
+/// A shell meta-sequence begins with a dollar sign character, `$` (U+0024), and
+/// is followed by one of:
+///
+/// - a parentheses-enclesed sequence, `(text)`, indicating a subshell command
+/// - a brace-enclosed sequence, `{text}`, indicating a variable expansion
+/// - a bare word, `text`, indicating a variable expansion
```
Comment (hawkw) [inline src/token.rs]: I think it might be worth putting a little design thought into whether or not we intend to reproduce all of the above syntax from POSIX sh. That said, I'm okay with merging this tokenizer now regardless.

File: src/token.rs [hunk 2]
```diff
@@ -279,6 +291,106 @@ pub fn keyval(text: &str) -> TokenResult<(&str, &str)> {
...
+    let mut rem = &text[1 ..];
+    //  Search the text for the end of the meta sequence. Large tokens (dquote,
+    //  squote, and shell_meta) jump the search.
+    //  This loop is finite: all the called tokenizers advance the text, and it
+    //  will eventually become empty and terminate as incomplete.
+    loop {
+        if rem.trim().is_empty() {
+            return Err(Err::Incomplete(Needed::Unknown));
+        }
+        if let Ok((rest, _)) = trim_left(dquote)(rem) {
```
Comment (hawkw) [inline src/token.rs]: TIOLI: It feels like there ought to be a way to make this less repetitive. What about something like
```rust
'outer: loop {
    // ...

    for tokenizer in &[dquote, squote, shell_meta] {
        if let Ok((rest, _)) = trim_left(tokenizer)(rem) {
            rem = rest;
            continue 'outer;
        }
    }

   // ...
}
```

Comment (hawkw) [reply src/token.rs]: I'll admit that it's possible that all I've done is make the code Worse. Your call. 🤷‍♀️

File: src/token.rs [hunk 3]
```diff
@@ -279,6 +291,106 @@ pub fn keyval(text: &str) -> TokenResult<(&str, &str)> {
...
+/// ```
+pub fn shell_meta(text: &str) -> TokenResult {
+    use nom::tag;
+    use nom::Err;
+    use nom::Needed;
+    let (text, _) = tag!(text, "$")?;
+    let close = match text.clone().chars().next() {
+        Some('(') => ')',
+        Some('{') => '}',
+        Some(_) => return word(text),
```
Comment (hawkw) [inline src/token.rs]: The number of early returns in this function make it a little hard to follow (IMO) --- it might be good if some of them had comments describing why we return in that case? Admittedly this one is fairly trivial, but still.

Comment (myrrlyn) [reply src/token.rs]: yeah, that's fair

File: src/token.rs (hunk 1 above)
Comment (myrrlyn) [reply src/token.rs]: These can always be changed as we figure out what the grammar should be; I just picked () and {} as enclosers to show I could.

As we define a better idea of the grammar, the tokenizers can follow suit

**Comment by myrrlyn** [comment]
Eliza I added more comments; I'm going to merge now though.

File: src/token.rs (hunk 1 above)
Comment (hawkw) [reply src/token.rs]: Yeah, I definitely didn't intend that comment as a blocker, just something I thought while reading the code. :)

**Comment by hawkw** [comment]
@myrrlyn 
> Eliza I added more comments; I'm going to merge now though.

if I give you a ✅, that typically means that I'm okay with the branch merging regardless of whether or not you make the changes I've suggested. :smile: 

Usually any comments I leave with an approved review are style nits or suggestions that are up to you if you want to take or not.

File: src/token.rs (hunk 2 above)
Comment (hawkw) [reply src/token.rs]: i think we could even do something like
```rust
if let Some((rest, _)) = &[dquote, squote, shell_meta] 
    .iter()
    .filter_map(|tokenizer| trim_left(tokenizer)(rem).ok())
    .next()
{
    rem = rest;
    continue;
}
```
which obviates the need to label the outer loop. but on the other hand, this may be less clear than the `for` loop. And, I suspect the compiler probably unrolls the loop into machine code that's very similar to the three `if`s, but I'm less sure about the iterators...

Comment (hawkw) [reply src/token.rs]: (sorry for overthinking this :P)

**Comment by myrrlyn** [comment]
I figured so lol hence the "I'm gonna do it just providing a courtesy update"
